wire-compiler/src/test/java/com/squareup/wire/schema/LinkerTest.kt

```diff
@@ -56,6 +59,272 @@ class LinkerTest {
     )
   }
 
+  @Test fun opaqueMessageDeclaredField() {
+    fs.add(
+      "source-path/cafe/cafe.proto",
+      """
+          |syntax = "proto2";
+          |
+          |package cafe;
+          |
+          |message CafeDrink {
+          |  optional int32 size_ounces = 1;
+          |  repeated EspressoShot shots = 2;
+          |}
+          |
+          |message EspressoShot {
+          |  optional Roast roast = 1;
+          |  optional bool decaf = 2;
+          |}
+          |
+          |enum Roast {
+          |  MEDIUM = 1;
+          |  DARK = 2;
+          |}
+      """.trimMargin(),
+    )
+    val schema = loadAndLinkSchema(opaqueTypes = listOf(ProtoType.get("cafe.EspressoShot")))
+    assertThat((schema.getType("cafe.CafeDrink") as MessageType).field("shots")!!.type!!)
+      .isEqualTo(ProtoType.BYTES)
+    assertThat(schema.protoFile("cafe/cafe.proto")!!.toSchema()).isEqualTo(
+      """
+        |// Proto schema formatted by Wire, do not edit.
+        |// Source: cafe/cafe.proto
+        |
+        |syntax = "proto2";
+        |
+        |package cafe;
+        |
+        |message CafeDrink {
+        |  optional int32 size_ounces = 1;
+        |
+        |  repeated bytes shots = 2;
+        |}
+        |
+        |message EspressoShot {
+        |  optional Roast roast = 1;
+        |
+        |  optional bool decaf = 2;
+        |}
+        |
+        |enum Roast {
+        |  MEDIUM = 1;
+        |  DARK = 2;
+        |}
+    |
+      """.trimMargin(),
+    )
+  }
+
+  @Test fun opaqueEnumDeclaredField() {
+    fs.add(
+      "source-path/cafe/cafe.proto",
+      """
+          |syntax = "proto2";
+          |
+          |package cafe;
+          |
+          |message CafeDrink {
+          |  optional int32 size_ounces = 1;
+          |  repeated EspressoShot shots = 2;
+          |}
+          |
+          |message EspressoShot {
+          |  optional Roast roast = 1;
+          |  optional bool decaf = 2;
+          |}
+          |
+          |enum Roast {
+          |  MEDIUM = 1;
+          |  DARK = 2;
+          |}
+      """.trimMargin(),
+    )
+    val schema = loadAndLinkSchema(opaqueTypes = listOf(ProtoType.get("cafe.Roast")))
+    assertThat((schema.getType("cafe.EspressoShot") as MessageType).field("roast")!!.type!!)
+      .isEqualTo(ProtoType.BYTES)
+    assertThat(schema.protoFile("cafe/cafe.proto")!!.toSchema()).isEqualTo(
+      """
+        |// Proto schema formatted by Wire, do not edit.
+        |// Source: cafe/cafe.proto
+        |
+        |syntax = "proto2";
+        |
+        |package cafe;
+        |
+        |message CafeDrink {
+        |  optional int32 size_ounces = 1;
+        |
+        |  repeated EspressoShot shots = 2;
+        |}
+        |
+        |message EspressoShot {
+        |  optional bytes roast = 1;
```

Comment:
I don’t think this should work this way. `bytes` are encoded as length-delimited, but enums are encoded using varints, and we can’t substitute one for the other while retaining the encoded form. We’d want enums to map to `int32`, or simply don’t support them.